### PR TITLE
Rewrite from header

### PR DESCRIPTION
**Pull request by marpio**
**Description**
This PR extends the SubjectAccessReviews rewrite functionality by allowing to rewrite the attributes using a value from http header.

**Motivation**
I would like to put kube-rbac-proxy in front of Loki (https://grafana.com/docs/loki/latest/). Loki, when deployed in multi-tenant mode, requires X-Scope-OrgID header containing the tenantID. In my case, the tenantID is just the namespace and so I could check if a given user/SA has access to a particular resource in a given namespace provided in the X-Scope-OrgID header. This way I could separate the logs on a per namespace basis.

### CODE REVIEW

**Comment by brancz** [comment]
cc @s-urbaniak @simonpasquier I don't have time to think through whether this is safe or not. At first glance it seems fine, but don't have the time to spend on this right now, so if you could have a look that would be amazing!

File: pkg/proxy/proxy.go
Comment (s-urbaniak) [inline pkg/proxy/proxy.go:160]: nit: i would continue returning `nil` here, there is no need to allocate an empty slice, as the zero value (`nil`) of a slice also returns 0 for its capacity and length.

File: pkg/proxy/proxy.go
Comment (s-urbaniak) [inline pkg/proxy/proxy.go:146]: suggestion: let's untangle those nested if/else statements, it starts being complex. We can have a more linear code if we change the method signature like so:

```
// GetRequestAttributes populates authorizer attributes for the requests to kube-rbac-proxy.
func (n krpAuthorizerAttributesGetter) GetRequestAttributes(u user.Info, r *http.Request) (allAttrs []authorizer.Attributes) {
	apiVerb := ""
	switch r.Method {
	case "POST":
		apiVerb = "create"
	case "GET":
		apiVerb = "get"
	case "PUT":
		apiVerb = "update"
	case "PATCH":
		apiVerb = "patch"
	case "DELETE":
		apiVerb = "delete"
	}

	defer func() {
		for attrs := range allAttrs {
			klog.V(5).Infof("kube-rbac-proxy request attributes: attrs=%#v", attrs)
		}
	}()

	if n.authzConfig.ResourceAttributes == nil {
		// Default attributes mirror the API attributes that would allow this access to kube-rbac-proxy
		allAttrs = append(allAttrs, authorizer.AttributesRecord{
			User:            u,
			Verb:            apiVerb,
			Namespace:       "",
			APIGroup:        "",
			APIVersion:      "",
			Resource:        "",
			Subresource:     "",
			Name:            "",
			ResourceRequest: false,
			Path:            r.URL.Path,
		})
		return
	}

	if n.authzConfig.Rewrites == nil {
		allAttrs = append(allAttrs, authorizer.AttributesRecord{
			User:            u,
			Verb:            apiVerb,
			Namespace:       n.authzConfig.ResourceAttributes.Namespace,
			APIGroup:        n.authzConfig.ResourceAttributes.APIGroup,
			APIVersion:      n.authzConfig.ResourceAttributes.APIVersion,
			Resource:        n.authzConfig.ResourceAttributes.Resource,
			Subresource:     n.authzConfig.ResourceAttributes.Subresource,
			Name:            n.authzConfig.ResourceAttributes.Name,
			ResourceRequest: true,
		})

		return
	}

	params := []string{}

	if n.authzConfig.Rewrites.ByQueryParameter != nil && n.authzConfig.Rewrites.ByQueryParameter.Name != "" {
		if ps, ok := r.URL.Query()[n.authzConfig.Rewrites.ByQueryParameter.Name]; ok {
			params = append(params, ps...)
		}
	}

	if n.authzConfig.Rewrites.ByHTTPHeader != nil && n.authzConfig.Rewrites.ByHTTPHeader.Name != "" {
		if p := r.Header.Get(n.authzConfig.Rewrites.ByHTTPHeader.Name); p != "" {
			params = append(params, p)
		}
	}

	if len(params) == 0 {
		return
	}

	for _, param := range params {
		allAttrs = append(allAttrs, authorizer.AttributesRecord{
			User:            u,
			Verb:            apiVerb,
			Namespace:       templateWithValue(n.authzConfig.ResourceAttributes.Namespace, param),
			APIGroup:        templateWithValue(n.authzConfig.ResourceAttributes.APIGroup, param),
			APIVersion:      templateWithValue(n.authzConfig.ResourceAttributes.APIVersion, param),
			Resource:        templateWithValue(n.authzConfig.ResourceAttributes.Resource, param),
			Subresource:     templateWithValue(n.authzConfig.ResourceAttributes.Subresource, param),
			Name:            templateWithValue(n.authzConfig.ResourceAttributes.Name, param),
			ResourceRequest: true,
		})
	}

	return
}
```

**Review by s-urbaniak — CHANGES_REQUESTED** [review]
thank you very much for the contribution! Generally this lgtm, i have just one suggestion to make the code linear.

Also, do you mind to add some unit tests?

Thank you again! :tada:

**Comment by marpio** [comment]
@s-urbaniak thanks for the review and suggestions!
I've refactored the func and added unit tests for it.
I had to adjust your proposal a bit b/c of some strange e2e tests failures (see i.e. https://github.com/brancz/kube-rbac-proxy/runs/1558365978)
I didn't have the time to investigate but the test were passing as soon as I removed the named return param.

File: pkg/proxy/proxy.go
Comment (s-urbaniak) [inline pkg/proxy/proxy.go:196]: nit: we can `return nil` here

**Comment by s-urbaniak** [comment]
@marpio yeah agreed, the non-named-return version looks better to me as well. I really have just one nit, else lgtm, thank you so much! :green_heart:

**Comment by marpio** [comment]
@s-urbaniak hmm.. returning nil causes the e2e test failure (see https://github.com/brancz/kube-rbac-proxy/pull/104/checks?check_run_id=1566450831)
I'm not sure what's the reason for that though...
Anyway returning `allAttrs` here sholdn't have any penalty b/c it's not initialized at this point.

**Comment by s-urbaniak** [comment]
@marpio interesting, let me inspect why that is the case :thinking:

**Comment by marpio** [comment]
@s-urbaniak have you been able to find anything?

**Comment by dbluxo** [comment]
Amazing @marpio this is exactly what we were looking for and already wanted to implement ourselves. Our usecase is the same as you describe to also enable the multi-tenancy capability with Grafana Loki. Would this configuration work after your PR is merged?:

```
"authorization":
  "resourceAttributes":
    "apiVersion": "metrics.k8s.io/v1beta1"
    "namespace": "{{ .Value }}"
    "resource": "pods"
  "rewrites":
    "byHTTPHeader":
      "name": "X-Scope-OrgID"
```

**Comment by s-urbaniak** [comment]
I think the failing e2e test was a flake as it passes for me locally with the `return nil` patch, if you want please submit a follow-up PR. However, this being a nit is not a stopper, hence merging. Sorry for the delay and thank you for the contribution!

**Comment by marpio** [comment]
@s-urbaniak thanks for taking the time!

**Comment by s-urbaniak** [comment]
@brancz I think this deservers a new release, however I have not the powers to publish new images. Happy to help out. wdyt?

**Comment by paulfantom** [comment]
Seems like after merging this, our e2e tests started failing :thinking: 

More in https://github.com/brancz/kube-rbac-proxy/runs/1748074270

Personally, I would like to know why those are failing before creating a release.

**Comment by s-urbaniak** [comment]
hmm, let's open an issue then, it seems to be related with certs rotation so i can have a look.

**Pull request merged**
